src/dispatcher/server.rs

```diff
@@ -148,40 +162,84 @@ impl SignatureDispatcher for Ed25519SignatureDispatcher {
 }
 
 #[tokio::main]
-async fn main() -> Result<(), Box<dyn std::error::Error>> {
+async fn main() -> remote_signer::Result<()> {
     SimpleLogger::from_env().init().unwrap();
     let config_arg = App::new("Remote Signer Dispatcher")
-        .arg(Arg::with_name("config")
-             .short("c")
-             .long("config")
-             .takes_value(true)
-             .value_name("FILE")
-             .default_value("dispatcher_config.json")
-             .help("Dispatcher .json configuration file")
-        ).get_matches();
+        .arg(
+            Arg::with_name("config")
+                .short("c")
+                .long("config")
+                .takes_value(true)
+                .value_name("FILE")
+                .default_value("dispatcher_config.json")
+                .help("Dispatcher .json configuration file"),
+        )
+        .get_matches();
+
+    info!("Start");
 
     let conf_path = config_arg.value_of("config").unwrap();
-    info!("Parsing configuration file `{}`.", conf_path);
-    let (config, keysigners) = config::parse_dispatcher(conf_path)?;
-    debug!("Parsed configuration file: {:?}", config);
-    let addr = config.bind_addr.parse()?;
-    let server_root_ca_cert = tokio::fs::read(&config.tlsauth.ca).await?;
-    let server_root_ca_cert = Certificate::from_pem(server_root_ca_cert);
-    let client_cert = tokio::fs::read(&config.tlsauth.client_cert).await?;
-    let client_key = tokio::fs::read(&config.tlsauth.client_key).await?;
-    let client_identity = Identity::from_pem(client_cert, client_key);
-    let tls_auth = ClientTlsConfig::new()
-        .ca_certificate(server_root_ca_cert)
-        .identity(client_identity);
-    let dispatcher = Ed25519SignatureDispatcher { config, tls_auth, keysigners };
+    let (addr, dispatcher) = create_dispatcher(conf_path)?;
     debug!("Initialized Dispatcher server: {:?}", dispatcher);
 
-    info!("Serving on {}...", addr);
+    let key_signers = Arc::clone(&dispatcher.keysigners);
 
-    Server::builder()
+    let mut server = Server::builder();
+    let serv = server
         .add_service(SignatureDispatcherServer::new(dispatcher))
         .serve(addr)
-        .await?;
+        .map_err(|err| RemoteSignerError::from(err));
+
+    let signal = reload_configs_upon_signal(&conf_path, key_signers);
+
+    info!("Serving on {}...", addr);
+    let result = future::try_join(serv, signal).await;
```

Comment:
matching `result` does not really add anything. You can use the `await?` syntax and wrap the empty value in an `Ok()` when it returns; like it was before pretty much.